snuba/clickhouse/http.py

```diff
@@ -57,8 +69,8 @@ def write(self, rows: Iterable[WriterTableRow]):
                 }
             ),
             headers={"Connection": "keep-alive", "Accept-Encoding": "gzip,deflate"},
-            body=map(self.__encode, rows),
-            chunked=True,
+            body=self._prepare_body(rows),
+            chunked=self.__chunk_size is None,
```

Comment:
There are two types of chunks, that's why I had an explicit chunked=True before. These are the scenarios:
- we pass chunked=True to urlopen and we pass an iterable where every string is an element (this is what happens today and should be preserved).
- we pass chunked = False to urlopen and we break the input into chunks according to the maximum size passed in
- we pass chunked = False to urlopen and we do not tweak the input, passing it as it is.

If I interpret chunked == None as "do not chunk" I still need a flag to tell urlopen to send a chunked request.

Reply:
nevermind, I removed the distinction.

Reply:
Discussed this in person before finishing this comment, but recording these notes in case we need to come back to it later.

`urllib3` will eventually call [one of these two methods](https://github.com/urllib3/urllib3/blob/5b047b645f5f93900d5e2fc31230848c25eb1f5f/src/urllib3/connectionpool.py#L376-L381) based on the `chunked` argument provided. When `not chunked`, we will [eventually end up here in the Python `http.client` module](https://github.com/python/cpython/blob/23669330b7d0d5ad1a9aac40315ba4c2e765f9dd/Lib/http/client.py#L1246-L1290)

The [important part of that method body](https://github.com/python/cpython/blob/23669330b7d0d5ad1a9aac40315ba4c2e765f9dd/Lib/http/client.py#L1257-L1262) for us:

```
# chunked encoding will happen if HTTP/1.1 is used and either
# the caller passes encode_chunked=True or the following
# conditions hold:
# 1. content-length has not been explicitly set
# 2. the body is a file or iterable, but not a str or bytes-like
# 3. Transfer-Encoding has NOT been explicitly set by the caller
```

Since the return type of `_prepare_body` is `Iterable[bytes]` types, we're hitting _all three conditions_ in every case. We're never not sending chunks, it's just a matter of how big the chunks are. It doesn't matter what we're passing for with `chunked`. This is also [documented here](https://docs.python.org/3/library/http.client.html#http.client.HTTPConnection.request) but it's sometimes hard to figure out where `urllib3` stops and `http.client` begins.

The way that I think this should work is that:

- If you want to send a chunked request, you should use a positive, nonzero integer value. This will result in the iterator being split up into chunks of that size, which are independently sent using `Transfer-Encoding: chunked`.
- If you don't want to send a chunked request, you should use `None`. This will result in the entire contents of the iterator until `StopIteration` being added to one big string that is used as the request body, and sent _without_ `Transfer-Encoding: chunked` and with the `Content-Length` header: https://github.com/python/cpython/blob/23669330b7d0d5ad1a9aac40315ba4c2e765f9dd/Lib/http/client.py#L1269-L1271